### ISSUE TITLE
Incorrect setReturnBuffers when parsing PubSub 'message' response

### ISSUE THREAD

**Issue by aleiby** [redis/node-redis#1870]
Since upgrading my redis server to 6.2.6 and redis client to 4.0.2 (from 4.0.1) I've been occasionally seeing the following crash:
`TypeError [ERR_INVALID_ARG_TYPE]: The "otherBuffer" argument must be an instance of Buffer or Uint8Array. Received type string ('message')`

I've tracked this down to `parser.optionReturnBuffers` being set to `false` when it should be `true` causing the 'message' string to be returned as a string rather than a Buffer.

[commands-queue.ts](https://github.com/redis/node-redis/blob/a2299509a0ac1ebabf486a0e4ad542fb087b7196/packages/client/lib/client/commands-queue.ts#L113) expects this to be 
a Buffer, and Buffer.equals throws the above NodeError.

This response comes in through [parseResponse](https://github.com/redis/node-redis/blob/a2299509a0ac1ebabf486a0e4ad542fb087b7196/packages/client/lib/client/commands-queue.ts#L370) which calls `this.#setReturnBuffers()`.  I haven't been able to track down why this works correctly most of the time, but occasionally winds up setting `parser.optionReturnBuffers` to `false` and blowing up.  Hopefully someone more familiar with the code can spot what's going on.

**Environment:**
 - **Node.js Version**: 16.8.0
 - **Redis Server Version**: 6.2.6
 - **Node Redis Version**: 4.0.2
 - **Platform**: Ubuntu 20.04.3 LTS (WSL2)

**Comment by aleiby** [comment]
Commenting out `setReturnBuffers` in `shiftWaitingForReply` avoids the crash for now.  More details in a comment I left here: https://github.com/redis/node-redis/commit/77022209bde327d6a2d004dd07153bb6140d0ae2

I expect that will cause other problems since that change was made for a reason (which was unfortunately not described in the commit message).

**Comment by leibale** [comment]
@aleiby I haven't debugged it yet, but I'm working on a new RESP parser, which (hopefully) will solve this problem as well. Will keep you posted.

**Comment by manikyafc** [comment]
This issue is affecting our production as well, when should we expect a fix for this?

**Comment by leibale** [comment]
@aleiby @manikyafc I think that the changes in #1899 solve this problem, do you have a reproduction of the issue?

**Comment by aleiby** [comment]
It happens fairly reliably in my codebase, but I was unable to distill it down into an easily sharable repro unfortunately.  You need to construct a scenario where `parseResponse` handles both a `subscribe` rely followed by a `message` reply in a single call.

**Comment by binarymist** [comment]
Same deal for us. This is the stack trace:

```
 node:buffer:816
     throw new ERR_INVALID_ARG_TYPE(
     ^

 TypeError [ERR_INVALID_ARG_TYPE]: The "otherBuffer" argument must be an instance of Buffer or Uint8Array. Received type string ('message')
     at new NodeError (node:internal/errors:371:5)
     at Buffer.equals (node:buffer:816:11)
     at JavascriptRedisParser.returnReply (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:42:123)
     at JavascriptRedisParser.execute (/usr/src/app/node_modules/redis-parser/lib/parser.js:544:14)
     at RedisCommandsQueue.parseResponse (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:194:71)
     at RedisSocket.<anonymous> (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/index.js:336:83)
     at RedisSocket.emit (node:events:390:28)
     at Socket.<anonymous> (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/socket.js:189:44)
     at Socket.emit (node:events:390:28)
     at addChunk (node:internal/streams/readable:324:12) {
   code: 'ERR_INVALID_ARG_TYPE'
 }
 
 Node.js v17.3.0

```

Call site: https://github.com/purpleteam-labs/purpleteam-orchestrator/blob/96fa97c7135130cd66d2d4a1e691f3cc84e32550/src/api/orchestration/subscribers/testerWatcher.js#L39

**Comment by binarymist** [comment]
@leibale : I tried your [branch](https://github.com/leibale/node-redis/commits/parser) as in:

Our package.json:
```
...
"redis": "leibale/node-redis.git#parser",
...
```

and we get:

```
Error [ERR_MODULE_NOT_FOUND]: Cannot find package '/usr/src/app/node_modules/redis/' imported from /our/file.js
     at new NodeError (node:internal/errors:371:5)
     at legacyMainResolve (node:internal/modules/esm/resolve:312:9)
     at packageResolve (node:internal/modules/esm/resolve:857:14)
     at moduleResolve (node:internal/modules/esm/resolve:910:18)
     at defaultResolve (node:internal/modules/esm/resolve:1005:11)
     at ESMLoader.resolve (node:internal/modules/esm/loader:530:30)
     at ESMLoader.getModuleJob (node:internal/modules/esm/loader:251:18)
     at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:79:40)
     at link (node:internal/modules/esm/module_job:78:36) {
   code: 'ERR_MODULE_NOT_FOUND'
 }
 
 Node.js v17.3.0
```

**Comment by leibale** [comment]
@binarymist first of all, thanks for helping! :)

Regarding the error, you can't `npm install` directly from GitHub, cause of typescript, which needs to be compiled/transpiled... You'll have to:

```bash
git clone --single-branch --branch parser https://github.com/leibale/node-redis
cd node-redis
npm install -ws
npm run build-all
```

then use [`npm link`](https://docs.npmjs.com/cli/v8/commands/npm-link) to link your local version

**Comment by NormandoHall** [comment]
Same here. Waiting the release that fixes this bug.

**Comment by binarymist** [comment]
Thanks @leibale.

Our project that consumes the redis NPM package that is experiencing this issue is run in a container.

I've added the following to the Dockerfile:

```
RUN cd /usr/src/ && \
  git clone --single-branch --branch parser https://github.com/leibale/node-redis && \
  cd node-redis && \
  npm install -ws && \
  npm run build-all && \
  npm link
```
package.json reverted to:

```
...
"dependencies": {
  ...
  "redis": "^4.0.2",
  ...
}
...
```

Still get the `ERR_MODULE_NOT_FOUND` error.

**Comment by binarymist** [comment]
@leibale: Which Pull Request is supposed to fix this, #1899 or the newer #2016 from @jhiesey?
Any further thoughts on how to test these changes in the container?
Is there an ETA on the fix? We currently have a [large release](https://github.com/purpleteam-labs/purpleteam/issues/107#issuecomment-1052917439) blocked on this.

**Comment by jhiesey** [comment]
@binarymist I can't speak for the maintainers, but my PR is a quick fix for the immediate issue. #1899 is a rewrite of the parser and is a better long term solution but I suspect it's not quite ready.

The workaround I'm currently using is to install [patch-package](https://www.npmjs.com/package/patch-package) and add this patch file at `patches/@node-redis+client+1.0.4.patch`:
```udiff
diff --git a/node_modules/@node-redis/client/dist/lib/client/commands-queue.js b/node_modules/@node-redis/client/dist/lib/client/commands-queue.js
index 8d22295..d3e7807 100644
--- a/node_modules/@node-redis/client/dist/lib/client/commands-queue.js
+++ b/node_modules/@node-redis/client/dist/lib/client/commands-queue.js
@@ -279,7 +279,8 @@ _a = RedisCommandsQueue, _RedisCommandsQueue_maxLength = new WeakMap(), _RedisCo
 }, _RedisCommandsQueue_setReturnBuffers = function _RedisCommandsQueue_setReturnBuffers() {
     var _b, _c;
     __classPrivateFieldGet(this, _RedisCommandsQueue_parser, "f").setReturnBuffers(!!((_b = __classPrivateFieldGet(this, _RedisCommandsQueue_waitingForReply, "f").head) === null || _b === void 0 ? void 0 : _b.value.returnBuffers) ||
-        !!((_c = __classPrivateFieldGet(this, _RedisCommandsQueue_pubSubState, "f")) === null || _c === void 0 ? void 0 : _c.subscribed));
+        !!((_c = __classPrivateFieldGet(this, _RedisCommandsQueue_pubSubState, "f")) === null || _c === void 0 ? void 0 : _c.subscribed) ||
+        !!((_c = __classPrivateFieldGet(this, _RedisCommandsQueue_pubSubState, "f")) === null || _c === void 0 ? void 0 : _c.subscribing));
 }, _RedisCommandsQueue_shiftWaitingForReply = function _RedisCommandsQueue_shiftWaitingForReply() {
     if (!__classPrivateFieldGet(this, _RedisCommandsQueue_waitingForReply, "f").length) {
         throw new Error('Got an unexpected reply from Redis');

```

**Comment by binarymist** [comment]
That looks helpful @jhiesey, although it seems to have introduced another issue:

```
TypeError: nonBlockingClients[channel].rpush is not a function
    at file:///usr/src/app/src/api/orchestration/subscribers/testerWatcher.js:72:41
    at Function._RedisCommandsQueue_emitPubSubMessage (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:231:9)
    at JavascriptRedisParser.returnReply (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:43:123)
    at JavascriptRedisParser.execute (/usr/src/app/node_modules/redis-parser/lib/parser.js:544:14)
    at RedisCommandsQueue.parseResponse (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/commands-queue.js:194:71)
    at RedisSocket.<anonymous> (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/index.js:336:83)
    at RedisSocket.emit (node:events:390:28)
    at Socket.<anonymous> (/usr/src/app/node_modules/@node-redis/client/dist/lib/client/socket.js:189:44)
    at Socket.emit (node:events:390:28)
    at addChunk (node:internal/streams/readable:324:12)
```
Which I'm now getting earler at: https://github.com/purpleteam-labs/purpleteam-orchestrator/blob/0f77b4a9d31844629f323891b5b9ce7f9e7ce98e/src/api/orchestration/subscribers/testerWatcher.js#L72

**Comment by binarymist** [comment]
Ok, the next problem was casing of `rpush`, found the answer many pages through the issues list [here](https://github.com/redis/node-redis/issues/1765).

Something is still broken though, I'm just not getting any errors now.

**Comment by leibale** [comment]
@binarymist change it to `nonBlockingClients[channel].rPush` (capital P)

**Comment by binarymist** [comment]
> @binarymist change it to nonBlockingClients[channel].rPush (capital P)

Yip, got that one thanks.

Something is still broken though, I'm going to have to spend more time debugging.

**Comment by digitalml** [comment]
I see this issue as closed but I am getting this error for pub sub messages randomly every now and then. "The "otherBuffer" argument must be an instance of Buffer or Uint8Array."  I've read through this closed issue but it's not clear to me How to resolve this? Can someone please point it out?

Thank you!

**Comment by mifopen** [comment]
ping

**Comment by mifopen** [comment]
@leibale Could you please explain why the issue was closed?

**Comment by mifopen** [comment]
Also, could anybody explain how such an exception could be caught? Currently, it's just crashing my server completely.

**Comment by jhiesey** [comment]
@digitalml @mifopen the issue is closed because the fix was merged to master. Unfortunately there hasn't been a release published that includes the fix; hopefully there will be one soon. Perhaps @leibale could comment on when.

You can use my workaround in https://github.com/redis/node-redis/issues/1870#issuecomment-1057530027 for the time being

**Comment by alexzaporozhets** [comment]
@leibale when do you plan to release the next version with this fix?

**Comment by leibale** [comment]
A new version with @jhiesey fix will be released tomorrow. Sorry it took so much time.

**Comment by leibale** [comment]
@alexzaporozhets @NormandoHall @digitalml @jhiesey @mifopen @binarymist 
version 4.0.6 is out! :) (do not try 4.0.5, I've made a mistake in the release process... :man_facepalming:)

**Comment by kvz** [comment]
I think a bug may still be present in @redis/client 5.10.0

While PR #1899 fixed the decoder to return Buffers, the handlers still lack defensive type checks and crash when strings slip through. When multiple responses arrive in one TCP chunk, `parser.optionReturnBuffers` gets out of sync. The decoder switches between buffer/string mode at the wrong moment.

This is why it's intermittent and hard to reproduce - requires specific timing where subscribe acknowledgment + first message arrive together.

### Reproduction

```js
Buffer.from('message').equals('message')
// TypeError [ERR_INVALID_ARG_TYPE]: The "otherBuffer" argument must be an instance of Buffer or Uint8Array
```

### Affected code in `pub-sub.js` (v5.10.0)

```js
// Line 27-32: isStatusReply()
COMMANDS[PUBSUB_TYPE.CHANNELS].subscribe.equals(reply[0])  // crashes if reply[0] is string

// Line 35: isShardedUnsubscribe()  
COMMANDS[PUBSUB_TYPE.SHARDED].unsubscribe.equals(reply[0])  // crashes if reply[0] is string

// Line 267-279: handleMessageReply()
if (COMMANDS[PUBSUB_TYPE.CHANNELS].message.equals(reply[0])) {  // crashes if reply[0] is string
```

### Suggested fix

Add type guard before each `.equals()` call:

```js
// Option A: Early return
if (!Buffer.isBuffer(reply[0])) return false;

// Option B: Safe helper
const safeEquals = (buf, val) => Buffer.isBuffer(val) ? buf.equals(val) : buf.toString() === val;
```

### Environment
- `redis@4.5.1` → `@redis/client@1.4.2` (affected)
- `redis@5.10.0` → `@redis/client@5.10.0` (still affected)
- Node.js v24.10.0, v22.x
- Intermittent crash in production pubsub handlers
